core/src/chain/header.rs

```diff
@@ -0,0 +1,276 @@
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash as _Hash, Hasher};
+
+use chain::entry::Entry;
+use chain::SourceChain;
+
+// @TODO - serialize properties as defined in HeadersEntrySchema from golang alpha 1
+// @see https://github.com/holochain/holochain-proto/blob/4d1b8c8a926e79dfe8deaa7d759f930b66a5314f/entry_headers.go#L7
+// @see https://github.com/holochain/holochain-rust/issues/75
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct Header {
+    /// the type of this entry
+    /// system types may have associated "subconscious" behavior
+    entry_type: String,
+    /// ISO8601 time stamp
+    time: String,
+    /// link to the immediately preceding header, None is valid only for genesis
+    next: Option<u64>,
+    /// mandatory link to the entry for this header
+    entry: u64,
+    /// link to the most recent header of the same type, None is valid only for the first of type
+    type_next: Option<u64>,
+    /// agent's cryptographic signature
+    signature: String,
+}
+
+impl _Hash for Header {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.entry_type.hash(state);
+        self.time.hash(state);
+        self.next.hash(state);
+        self.entry.hash(state);
+        self.type_next.hash(state);
+        self.signature.hash(state);
+    }
+}
```

Comment:
I believe, if you just need this you can just `#[derive(Hash)]` for header.
https://doc.rust-lang.org/std/hash/index.html

Reply:
And you don't even need to import `Hash` or `Hasher` for this.

Reply:
yup, whether or not we use the native hashing trait, there's already a ticket up to review all of this and swap it out for our hashing algo

https://github.com/holochain/holochain-rust/issues/20

the point here is just to get tests in place around the _intent_ of hashes (e.g. this hash points to that thing, or changing this property changes the hash)

and i don't want to just derive Hash because:
- need a network friendly hashing algo that works on a network with non-rust hc
- i might want to put properties on entries for convenience that don't end up in the hash/serialized form (e.g. https://github.com/holochain/holochain-rust/issues/73)
- we may also move away from implementing the native hashing and do our own thing anyway